pkg/tests/end_to_end_tests/ha_single_promscale_test.go

```diff
@@ -0,0 +1,579 @@
+		}
+		defer ing.Close()
+		for _, step := range testCase.steps {
+			t.Log(step.desc)
+			samples := generateHASamples(step.input)
+			rows, err := ing.Ingest(samples, ingestor.NewWriteRequest())
+			t.Logf("Num rows ingested: %d\n", rows)
+			if err != nil {
+				t.Fatalf("unexpected error: %v", err)
+				return
+			}
+			if !checkDbState(t, db, step.output) {
+				return
+			}
+
+			if step.tickSyncRoutine {
+				ticker.Tick()
+				// sleep this routine so the sync routine has had time
+				// to be scheduled and read the ticker
+				// and the try_change_leader routine has also been
+				// scheduled
+				time.Sleep(time.Second)
+			}
+
+			if err := setLeaseExplicitly(db, step.explicitLease); err != nil {
+				t.Fatalf("could not change lease state from outside: %v", err)
+				return
+			}
+
+		}
+	})
+}
+
+func prepareIngestorWithHa(db *pgxpool.Pool, t testing.TB) (*manualTicker, *ingestor.DBIngestor, error) {
+	// manuel ticker, ticked when we want
+	// to explicitly let the state sync routine run
+	ticker := &manualTicker{c: make(chan time.Time, 1)}
+	// function that returns time that is always in the future
+	// so the calls to ha.Service.checkLeaseUntilAndLastWrite
+	// depends only on the max time seen by an instance
+	// and try_change_leader is called when we want
+	tooFarInTheFutureNowFn := func() time.Time { return time.Now().Add(time.Hour) }
+	haService := ha.NewHAServiceWith(pgxconn.NewPgxConn(db), ticker, tooFarInTheFutureNowFn)
+	haParser := ha.NewHAParser(haService)
+	cach := &cache.MetricNameCache{Metrics: clockcache.WithMax(cache.DefaultMetricCacheSize)}
+	ing, err := ingestor.NewPgxIngestor(pgxconn.NewPgxConn(db), cach, haParser, &ingestor.Cfg{})
+	if err != nil {
+		t.Fatalf("could not create ingestor: %v", err)
+	}
+	return ticker, ing, err
+}
+
+func TestHALeaderChangeDueToInactivity(t *testing.T) {
+	testCase := haTestCase{
+		db: "ha_leader_change_due_to_inactivity",
+		steps: []haTestCaseStep{
+			{
+				desc: "1 becomes leader",
+				input: haTestInput{
+					replica: "1",
+					minT:    time.Unix(0, 0),
+					maxT:    time.Unix(0, 0),
+				},
+				output: haTestOutput{
+					expectedNumRowsInDb: 1,
+					expectedMaxTimeInDb: time.Unix(0, 0),
+					expectedLeaseStateInDb: leaseState{
+						cluster:    "cluster",
+						leader:     "1",
+						leaseStart: time.Unix(0, 0),
+						leaseUntil: time.Unix(60, 0),
+					},
+				},
+				tickSyncRoutine: true,
+			}, {
+				desc: "2 sends data in lease, no insert",
+				input: haTestInput{
+					replica: "2",
+					minT:    time.Unix(0, 0),
+					maxT:    time.Unix(0, 0),
+				},
+				output: haTestOutput{
+					expectedNumRowsInDb: 1,
+					expectedMaxTimeInDb: time.Unix(0, 0),
+					expectedLeaseStateInDb: leaseState{
+						cluster:    "cluster",
+						leader:     "1",
+						leaseStart: time.Unix(0, 0),
+						leaseUntil: time.Unix(60, 0),
+					},
+				},
+				tickSyncRoutine: true,
+			}, {
+				desc: "2 sends data after lease, no insert -> triggers leader change",
```

Comment:
This isn't great. I shouldTryToChangeLeader and the new lease will include this point. But, it's not written because it needs to trigger a tick first. Does this mean we will loose some points for any leader change? Should we introduce a non-async code path for try leader change here to avoid this?

Reply:
Yes, points will be lost until the sync routine ticks and updates the lease. Latest commit will sync with the db to see if leader changed from outside, but will not try to change leader